protocols/kad/src/async_behaviour.rs

```diff
@@ -0,0 +1,346 @@
+use std::{collections::HashMap, task::Poll};
+
+use futures::channel::mpsc;
+use libp2p_core::{Endpoint, Multiaddr};
+use libp2p_identity::PeerId;
+use libp2p_swarm::{
+    ConnectionDenied, ConnectionId, FromSwarm, NetworkBehaviour, THandler, THandlerInEvent,
+    THandlerOutEvent, ToSwarm,
+};
+
+use crate::{
+    kbucket, store, AddProviderResult, Behaviour, BootstrapResult, Event, GetClosestPeersResult,
+    GetProvidersResult, GetRecordResult, NoKnownPeers, PutRecordResult, QueryId, QueryResult,
+    QueryStats, Quorum, Record, RecordKey,
+};
+
+/// The results of Kademlia queries (strongly typed) and only
+/// those initiated by the user.
+pub struct AsyncQueryResult<T> {
+    pub id: QueryId,
+    pub result: T,
+    pub stats: QueryStats,
+}
+impl<T> AsyncQueryResult<T> {
+    fn map<Out>(self, f: impl Fn(T) -> Out) -> AsyncQueryResult<Out> {
+        AsyncQueryResult {
+            id: self.id,
+            stats: self.stats,
+            result: f(self.result),
+        }
+    }
+}
+
+type UnboundedQueryResultSender<T> = mpsc::UnboundedSender<AsyncQueryResult<T>>;
+pub type UnboundedQueryResultReceiver<T> = mpsc::UnboundedReceiver<AsyncQueryResult<T>>;
+
+enum QueryResultSender {
+    Bootstrap(UnboundedQueryResultSender<BootstrapResult>),
+    GetClosestPeers(UnboundedQueryResultSender<GetClosestPeersResult>),
+    GetProviders(UnboundedQueryResultSender<GetProvidersResult>),
+    StartProviding(UnboundedQueryResultSender<AddProviderResult>),
+    GetRecord(UnboundedQueryResultSender<GetRecordResult>),
+    PutRecord(UnboundedQueryResultSender<PutRecordResult>),
+}
+
+/// A wrapper of [`Behaviour`] allowing to easily track
+/// [`QueryResult`] of user initiated Kademlia queries.
+///
+/// For each queries like [`Behaviour::bootstrap`], [`Behaviour::get_closest_peers`], etc
+/// a corresponding method ([`AsyncBehaviour::bootstrap_async`], [`AsyncBehaviour::get_closest_peers_async`])
+/// is available, allowing the developer to be notified from a typed [`UnboundedQueryResultReceiver`]
+/// instead from the normal [`Event::OutboundQueryProgressed`] event.
+///
+/// If a [`QueryResult`] has no corresponding [`UnboundedQueryResultReceiver`] or
+/// if the corresponding one has been dropped, it will simply be forwarded to the `Swarm`
+/// with an [`Event::OutboundQueryProgressed`] like nothing happen.
+///
+/// For more information, see [`Behaviour`].
+pub struct AsyncBehaviour<TStore> {
```

Comment:
Yeah it would probably be great !! 

I'm thinking about it but I really don't see how I could implement it since I need to capture the events emitted by the `kad::Behaviour`. Do you have an idea ?

Reply:
@jxs , I've thought a little bit about this. The only way I see to make a `Control` work for the `kad::Behaviour` is to have the entire `Behaviour` behind an `Arc<Mutex<>>`. I don't know if it is a good or bad idea. 

Will the "control" semantic become something standard for all the behaviours ? If so why not. But if not, I don't know how it will be understood by the end user to have some behaviours using a `control` and others not.

I'd really like to also have your opinion on this @guillaumemichel when you have the time.

Reply:
> @jxs , I've thought a little bit about this. The only way I see to make a Control work for the kad::Behaviour is to have the entire Behaviour behind an Arc<Mutex<>>. I don't know if it is a good or bad idea.

I don't think you need an Arc/ Mutex.
You could split the current `AsyncBehavior` logic into two parts: the Control part and the NetworkBehavior part, that communicate through an additional mpsc channel (the "command-channel").

- `Control` would have all the async variants of the kademlia behavior (`get_closest_peers_async`, `get_providers_async`, etc.).
There, it would create the mpsc channel for the results, and then send the a tuple (`Command::GetProviders {..}, mpsc::Sender<AsyncQueryResult>)` through the "command-channel" to the `AsyncBehavior`.
- `AsyncBehavior` would in its `poll_next` loop poll the receiving side of the command-channel, and handle the incoming command by calling the matching function on the wrapped kademlia behavior and storing the `Sender` in the `query_result_senders` hashmap. `poll_next` would also still have the already existing logic for intercepting `ToSwarm::GenerateEvent` events from the inner behavior and forwarding the results.

The only thing that we'd lose would be the deref to the inner Kademlia behavior, so we'd need to do the above sequence with the channel even for non-async methods like `add_address`.

Does that make sense? @jxs was that roughly what you had in mind?